lymph/unilateral.py

```diff
@@ -940,44 +958,13 @@ def check_and_assign(self, new_params: np.ndarray):
             :class:`Marginalizor`) all raise a ``ValueError`` when provided with
             invalid parameters.
         """
-        k = len(self.spread_probs)
-        if self.states == 2:
-            new_spread_probs = new_params[:k]
-            new_marg_params = new_params[k:]
-        if self.states == 3:
-            new_microscopic_parameter = new_params[0]
-            new_growth_probability = new_params[1]
-            new_spread_probs = new_params[2:k+2]
-            new_marg_params = new_params[k+2:]
-            if new_microscopic_parameter < 0. or new_microscopic_parameter > 1:
-                raise ValueError(
-                    "microscopic scaling parameter must be above 0 and below 1"
-                )
-            self.microscopic_parameter = new_microscopic_parameter
 
-            if new_growth_probability < 0. or new_growth_probability > 1.:
-                raise ValueError(
-                    "growth probability must be between 0 and 1"
-                )
-            self.growth_probability = new_growth_probability
-
-        try:
-            self.diag_time_dists.update(new_marg_params)
-        except ValueError as val_err:
-            raise ValueError(
-                "Parameters for marginalization over diagnose times are invalid"
-            ) from val_err
-
-        if new_spread_probs.shape != self.spread_probs.shape:
-            raise ValueError(
-                "Shape of provided spread parameters does not match network"
-            )
-        if np.any(0. > new_spread_probs) or np.any(new_spread_probs > 1.):
-            raise ValueError(
-                "Spread probs must be between 0 and 1"
-            )
+        for key, value in kwargs.items():
+            if key[:5] == 'micro':
+                self.parameters[key].micro_mod = value
+            else:
+                self.parameters[key].spread_prob = value

```

Comment:
I think this is definitely a better approach. I don''t like the solution at this point either...
I will also implement this change 👍🏽

Reply:
What we could also do is something like this:

```python
def set_spread_prob(mew_spread_prob: float):
    if 0. <= new_spread_prob <= 1.:
        raise ValueError(...)
    self._spread_prob = new_spread_prob

@spread_prob.setter
def spread_prob(self, new_spread_prob) -> float:
    self.set_spread_prob(new_spread_prob)
```
Which would combine both approaches. The other way around would also be possible:
```python
def set_spread_prob(mew_spread_prob: float):
    self.spread_prob = new_spread_prob

@spread_prob.setter
def spread_prob(self, new_spread_prob) -> float:
    if 0. <= new_spread_prob <= 1.:
        raise ValueError(...)
    self._spread_prob = new_spread_prob
```
Then the lookup dictionary can be assigned the function: `self._edge_lookup[edge.name] = edge.set_spread_prob`

Reply:
Just out of curiosity, what speaks against using:
```
def set_spread_prob(self, new_spread_prob: float) -> None:
        """Set the spread probability of the edge."""
        new_spread_prob = float(new_spread_prob)

        if not (0. <= new_spread_prob <= 1.):
            raise ValueError("Spread probability must be between 0 and 1!")

        self.spread_prob = new_spread_prob
```

Reply:
Because in that case, `spread_prob` would be considered a public attribute that a user should be allowed to modify themselves. That in turn could cause problems, because it would circumvent the checks. By encapsulating this attribute with getter and setter methods that work on an attribute `_spread_prob`, we indicate to the user that they should not interact with this directly, but only through the user-facing methods.

Of course, still nothing stops the user (or us) from modifying `_spread_probs` directly. In Python, unlike in languages like Java, this isn't forbidden. But it's a convention to regard everything starting with an underscore as private.

Reply:
Got it 👍🏽

Reply:
What I thought was that we might want to have _both_ the possibility to set the spread probabilities via
```python
edge.spread_prob = 0.45
```
_and_ pass the setter method to the edge lookup dictionary:
```python
self._edge_lookup[edge.name] = edge.set_spread_prob
```
I guess it'd be fine if we don't implement the setter and only allow setting via the explicit method, but it also wouldn't hurt.

However, what's not possible is this:
```python
@sprad_prob.setter
def set_spread_prob(self, new_spread_prob):
    ...
```
For two reasons:
1. The setter method must have the same name as the getter for the `@property` stuff to work
2. A method that has been decorated with such a setter cannot be "called" anymore. So, you can't pass it to the edge lookup dictionary anymore.

Reply:
I discovered an even better method! 🥳

The `@property` decorator can also be used like a normal function ([see here for documentation](https://docs.python.org/3/library/functions.html#property)):
```python
def get_spread_prob(self) -> float:
    """Return the spread probability."""
    return self._spread_prob

def set_spread_prob(self, new_spread_prob):
    """Set the spread probability."""
    # do some checks first...
    self._spread_prob = new_spread_prob

spread_probs = property(get_spread_probs, set_spread_probs, doc="The spread probability")
```
That way, the getter and setter methods remain available and we don't have to implement anything twice.

Reply:
I used this implementation now. It works but has one little issue (in my optinion).

The getter needs to be called as :
```
edge.get_spread_prob()
```
without the `()` only the function object will be returned instead of the value.